### PR TITLE
trace_callMany: tests and not loosing `ibs` state of all executed txs

### LINKED ISSUE #11798: eirgon3 trace_callMany not work with multiple trasaction
#### System information

Erigon version: `./erigon3 -- main branch [2fceaea]`

OS & Version: Linux

Commit hash: 2fceaea

Erigon Command (with flags/config):

datadir : '/mnt/evm_node/data/mainnet'
externalcl : true
prune.mode : 'archive'
prune.distance : 2048
prune.distance.blocks : 2048
chain : 'mainnet'
port : 30303
http.port : 8545
http.addr : 0.0.0.0
authrpc.port : 8551
torrent.port : 42069
http.vhosts : 'any'
http.corsdomain : '*'
private.api.addr : 127.0.0.1:9090
rpc.batch.limit : 200
http : true
ws : true
ws.compression : true
ws.port : 8546
http.api : ["eth","debug","net","trace","web3","erigon","admin"]
torrent.download.rate : '512mb'

Consensus Layer:  prysm

Consensus Layer Command (with flags/config):

Chain/Network:  mainnet

#### Expected behaviour
trace_callMany execute transaction on the state of pre-transaction

#### Actual behaviour
trace_callMany execute transaction seperately

#### Steps to reproduce the behaviour
execute the following code:

```
const simulate_address = "0x***57c3";
const weth_holder = "0x***7806";

const transfer_weth_txn = {
    from: weth_holder,
    to: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
    data: '0xa9059cbb' + simulate_address.slice(2).padStart(64, '0') + BigInt(1e20).toString(16).padStart(64, '0'),
}

const balanceOf_weth_txn = {
    from: simulate_address,
    to: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
    data: '0x70a08231' + simulate_address.slice(2).padStart(64, '0'),
}

async function test() {
    const txns = [
        [balanceOf_weth_txn, ['trace']],
        [transfer_weth_txn, ['trace']],
        [balanceOf_weth_txn, ['trace']]
    ];

    const blockNumber = 20635100;
    const blockNumberHex = '0x' + blockNumber.toString(16);

    const requestBody = [{
        "jsonrpc": "2.0",
        "method": "trace_callMany",
        "params": [
            txns,
            blockNumberHex
        ],
        "id": 0
    }];

    console.log('simulate_http_url:', hre.network.config.simulate_http_url);
    const res = await fetch(hre.network.config.simulate_http_url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'header': 'content-type: application/json'
        },
        body: JSON.stringify(requestBody),
    });

    const resJson = await res.json();

    for (const r of resJson[0].result) {
        console.log('r:', r);
    }
}

```
test().then(() => {
    console.log('test finished');
}).catch(err => {
    console.log('test failed:', err);
});

it trace_callMany with 3 transactions: [balanceOf, transfer, balanceOf ]

the third transaction's output should reflect the transfer result as '0x0000000000000000000000000000000000000000000000056bc75e2d63100000' 

while erigon3 return 
'0x0000000000000000000000000000000000000000000000000000000000000000'

works well with erigon2.

### PR DESCRIPTION
**Pull request by JkLondon**
closes #11798

### CODE REVIEW

File: turbo/jsonrpc/trace_adhoc.go [hunk 1]
```diff
@@ -1356,6 +1357,8 @@ func (api *TraceAPIImpl) doCallMany(ctx context.Context, dbtx kv.Tx, msgs []type
 		}
 	}
 
+	ibs.Reset()
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/trace_adhoc.go:1360]: feels weird that we return `ibs` object which was reset. who needs an empty object.

File: turbo/jsonrpc/trace_adhoc.go [hunk 2]
```diff
@@ -1258,7 +1261,7 @@ func (api *TraceAPIImpl) doCallMany(ctx context.Context, dbtx kv.Tx, msgs []type
 			vmConfig.Tracer = &ot
 		}
 
-		blockCtx := transactions.NewEVMBlockContext(engine, header, parentNrOrHash.RequireCanonical, dbtx, api._blockReader, chainConfig)
+		blockCtx = transactions.NewEVMBlockContext(engine, header, parentNrOrHash.RequireCanonical, dbtx, api._blockReader, chainConfig)
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/trace_adhoc.go:1264]: move it before for loop

File: turbo/jsonrpc/trace_adhoc.go [hunk 3]
```diff
@@ -1286,8 +1289,6 @@ func (api *TraceAPIImpl) doCallMany(ctx context.Context, dbtx kv.Tx, msgs []type
 			finalizeTxStateWriter = noop
 		}
 
-		ibs.Reset()
```
Comment (AskAlexSharov) [inline turbo/jsonrpc/trace_adhoc.go:1289]: Removal of `Reset` here - probably will change `sd.CompareStates(initialIbs, ibs)`.
Example if you send 2 txs. 1st txn sends coin from A to B and from C to D, 2nd txn sends coin from B to A. Then state diff of 2nd txn may include `C to D` event too (even that 2nd txn didn't touch C and D accounts). 
Please add this case to unit-test.

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

**Review by AskAlexSharov — APPROVED** [review]

File: turbo/jsonrpc/trace_adhoc.go (hunk 3 above)
Comment (JkLondon) [reply turbo/jsonrpc/trace_adhoc.go:1289]: added test, check pls

**Pull request merged**